File: src/lib/vision-context.ts
/**
 * @fileOverview Vision Context helpers that annotate perception for the cortex.
 * Maintains a rolling window of interpreted vision events to provide situational awareness.
 * Ported from Vision Context Python module.
 */

export interface VisionEvent {
  description: string;
  intent: string;
  confidence: number;
  timestamp: string;
}

class VisionContext {
  private events: VisionEvent[] = [];
  private readonly span: number;

  constructor(span: number = 5) {
    this.span = span;
  }

  /**
   * Pushes a new vision event into the rolling window.
   */
  push(description: string, intent: string, confidence: number): void {
    this.events.push({ 
      description, 
      intent, 
      confidence,
      timestamp: new Date().toISOString()
    });
    
    // Maintain rolling window (maxlen=span)
    if (this.events.length > this.span) {
      this.events.shift();
    }
  }

  /**
   * Returns a snapshot of current vision events for cortical processing.
   */
  snapshot(): { events: VisionEvent[]; count: number } {
    return {
      events: [...this.events],
      count: this.events.length
    };
  }

  /**
   * Clears all vision events.
   */
  clear(): void {
    this.events = [];
  }
}

// Singleton instance for global access
export const visionContext = new VisionContext();

File: src/vision_context.py
"""Vision context helpers that annotate perception for the cortex."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from typing import Deque, Dict, Any


@dataclass
class VisionEvent:
    description: str
    intent: str
    confidence: float


class VisionContext:
    """Maintains a rolling window of interpreted vision events."""

    def __init__(self, span: int = 5) -> None:
        self._events: Deque[VisionEvent] = deque(maxlen=span)

    def push(self, description: str, intent: str, confidence: float) -> None:
        self._events.append(VisionEvent(description, intent, confidence))

    def snapshot(self) -> Dict[str, Any]:
        return {
            "events": [event.__dict__ for event in list(self._events)],
            "count": len(self._events),
        }


_context: VisionContext | None = None


def get_context() -> VisionContext:
    global _context
    if _context is None:
        _context = VisionContext()
    return _context
